### ISSUE TITLE
Add End-of-life (EOL) / End-of-support (EOS) information for components

### ISSUE THREAD

**Issue by hvardhan20** [DependencyTrack/dependency-track#3108]
### Current Behavior

Currently, Dtrack identifies Out-of-date components by checking for newer versions of components. But it does NOT identify whether a component is Out-of-Support or not.

### Proposed Behavior

We should be able to get EOL/EOS information of components. 

Use case:
Take Spring boot for example. Couple of versions(2.7, 3.0) of Spring boot are nearing End of support date(for OSS). 
Support timelines of recent Spring boot versions from https://spring.io/projects/spring-boot#support:
<html>
<body>

Branch | Initial Release | End of Support | End Commercial Support *
-- | -- | -- | --
3.2.x | 2023-11-23 | 2024-11-23 | 2026-02-23
3.1.x | 2023-05-18 | 2024-05-18 | 2025-08-18
3.0.x | 2022-11-24 | 2023-11-24 | 2025-02-24
2.7.x | 2022-05-19 | 2023-11-24 | 2025-08-24
2.6.x | 2021-11-17 | 2022-11-24 | 2024-02-24

</body>
</html>

Which means there will not be a patch available if a critical security vulnerability is identified. Given how widely Spring boot is used globally, there would a security crisis if another Log4j scenario occurs. 

EOL/EOS feature would be a proactive step towards mitigating another Log4j. 

This information is tracked by https://endoflife.date/
We could add an analyzer for https://endoflife.date/ to fetch this information for components that are tracked currently.

### Checklist

- [X] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/master/CONTRIBUTING.md#filing-issues)
- [X] I have checked the [existing issues](https://github.com/DependencyTrack/dependency-track/issues) for whether this enhancement was already requested

**Comment by CorbyAsCode** [comment]
+1

I've always found it to be a good practice to provide viable and safe options instead of saying "Don't do this".  Development teams will be searching for alternatives for packages if the ones they're using are flagged for vulnerabilities.  Why not provide at least the latest version that's not vulnerable and isn't nearing end of support?  At best, we could provide a list of package versions that pass vuln tests.

**Comment by nscuro** [comment]
We've been waiting for endoflife.date to support PURL, which [as far as I can tell](https://endoflife.date/docs/api) is still not the case. Without PURL support, we can't reliably correlate packages we ingest from SBOMs with the EOL data provided by endoflife.date.

**Comment by setchy** [comment]
+1 - this would be a great enhancement to Dependency Track, especially when paired with Policy Management to define organizational operational/security policies

**Comment by setchy** [comment]
Found this issue which talks about PURL support for endoflife.date - https://github.com/endoflife-date/release-data/issues/51

**Comment by jimklimov** [comment]
Slightly agreeing with #8 that using not-outdated components may be safer than using those on life support (LTS dependencies). Makes sense.

On the other hand, code that was out in the wild for months or years and only collected a few known vulnerabilities (and one-off backports to fix them) may have something to it, compared to new experimental code that has frankly an unknown attack surface. For teams that deal with their own LTS product support, the presumed lack of API breaking changes coming via bug-fixes to LTS streams of their dependencies is a greater benefit than having some all-new code complete with new APIs and new bugs.

After all, the development tips and recent rolling releases of products may depend on "everything new", while "stable/LTS" releases may have to depend on something *predictable*, prioritizing for minimum human work (and minimal end-user surprise) needed to make it secure vs. problems that become known during the LTS timeframe.

**Comment by ptdropper** [comment]
I agree with @jimklimov . Latest and most current software version does not at all mean most secure. We know that as a fact in software. For security libraries, such as openssl, there are defined branches of that library that provide support for specific classes of hardware and operating system combinations. Older branches have fewer features thus a smaller attack surface, thus in my world those are MORE secure.

**Comment by jamietanna** [comment]
FYI I've built functionality in [a similar project](https://dmd.tanna.dev) for this.

I'd be happy to chat through the pURL-to-package name lookup, the code I've got is Apache-2.0 licensed.

**Comment by stevespringett** [comment]
Xeol now as an API, which would be ideal for Dependency-Track to integrate with.

**Comment by azabujuban** [comment]
Is there any standard field in SBOM that can be used to keep the EOL info? The only thing i can think of is the `properties` field, but it's not standard no?

**Comment by Najafov007** [comment]
> Xeol now as an API, which would be ideal for Dependency-Track to integrate with.

But until then, there's no chance of getting any info about EOL of our components as I've searched the whole internet, right sir?

**Comment by stevespringett** [comment]
> Is there any standard field in SBOM that can be used to keep the EOL info? The only thing i can think of is the properties field, but it's not standard no?

No. The EOL/EOS data is too dynamic and subject to change often. Its also prone to change due to market conditions where something that was previously unsupported, suddenly becomes supported for a while again, and eventually retired again. Happens all the time with open source and commercial software.

The Common Lifecycle Enumeration (CLE) project is looking into devising a time-series specification to represent milestone events  as well as other things that happen frequently in the supply chain, such as products being renamed/rebranded, mergers and acquisitions, etc. If anyone wants to get involved in that effort, there's a dedicated Slack channel on the OWASP Slack workspace and regularly scheduled working group calls.

One major issue with endoflife.date and similar efforts is that they are not time-series. They simply specify dates for various things and that's all. Product lifecycles are much more fluid than that, so we're trying to capture events in a time-series way. 

Once CLE is complete it will be incorporated into the [Transparency Exchange API](https://tc54.org/tea/) being co-developed by OWASP and Ecma TC54. The API will eventually be added to Dependency-Track which would make consuming SBOMs and other supply chain intelligence much more scalable than it is today. 

That said, there likely is a need for a short-term solution. Once that can work for the limited use cases we have today, while still being designed in a way that will work with these more robust future efforts.

**Comment by moulsonp** [comment]
Agree with @stevespringett it doesn't make sense to add this in the SBOM itself as it is too dynamic. But there is definately a need and demand for a short-term solution.

Can an attribute be added within dependency track where this EoS/EoES information can at least be manually added, ideally with an option to attempt to retreive via the endoflife.date API.

**Comment by sfl0r3nz05** [comment]
Hi, I don't know if there are plans to work on a short-term solution, given the demand:

- From DT point of view, it would be interesting to include an EOL policy enforcement approach, e.g.: the one followed by Sonatype.

- This could be endorsed with the integration of EOL in the SBOM as indicated in the [cdx:lifecycle Namespace Taxonomy](https://github.com/CycloneDX/cyclonedx-property-taxonomy/blob/479acf89cc0595995ad20d7fa9439f007250b050/cdx/lifecycle.md). E.g:

```json
"component" : {
    "name" : "[PRODUCT_NAME]",
    "version" : "2.2.3".
    "type" : "application",
    "bom-ref" : "dd8fc70b-767a-4398-885c-bbd0e8f6c68",
    "properties" : [
        {
            "name" : "cdx:lifecycle:milestone:endOfSupport",
            "value" : "2026-02-07T22:00:0Z"
        },
        {{
            "name" : "medcrypt:lifecycle:milestone:levelOfSupportText",
            "value" : "Q1 2026"
        }
```

**Comment by dcentrica** [comment]
For those that wish to use DependencyTrack and avail themselves of EOL data, we built [Metaport](https://gitlab.com/dcentrica/metaport/metaport-server) which integrates with DependencyTrack. It generates lifecycle graphics which can be shared with stakeholders to facilitate future planning and budgeting.

The project is currently in beta, so feedback, bugs, and all the rest are welcome.

<img width="1457" height="651" alt="Image" src="https://github.com/user-attachments/assets/acd4a723-4974-462c-8b0f-df7952ecb508" />

**Comment by msymons** [comment]
CLE V1.0 was [published](https://ecma-international.org/publications-and-standards/standards/ecma-428/) a few weeks ago. The full specification can downloaded from this page.  It makes for easy reading and includes a list of a few items that did not make it into this version, such as component forking and export restrictions.

As stated above, DT will not support CLE until the Transparency Exchange API (TEA) Specification is released.  TEA is currently scheduled to be submitted for approval by the Ecma General Assembly in June 2026.  So, we are definitely getting closer!

**Comment by Chriz76** [comment]
> We've been waiting for endoflife.date to support PURL, which [as far as I can tell](https://endoflife.date/docs/api) is still not the case. Without PURL support, we can't reliably correlate packages we ingest from SBOMs with the EOL data provided by endoflife.date.

Looks like endoflife API v1 now supports PURLs: https://endoflife.date/docs/api/v1/#/Identifiers/identifier_by_type

**Comment by Chriz76** [comment]
Would it be possible to use endoflife data as pseudo vulnerabilities in a Private Vulnerability Repo? 

Or would it be better to Dependency-Track webhook that queries an EOL API whenever a new component is ingested or updated?

**Comment by nscuro** [comment]
I think it would make the most sense to track this data separately from vulnerabilities, and then letting users access it through policies. 

Ideally you'd then be able to not only flag when a component is already EOL, but also when its EoL date is in the next N days or weeks.

**Comment by ptdropper** [comment]
I agree with @nscuro and can add that some EOS dates get changed over time. 
OS can be changed by purchasing extended support for the users of the product. 
EOS for some products have been changed by the manufacturer such that what was not supported at one time, the choice was reversed, and then is back to supported. 
EOS should stay as a separate characteristic.
I would use policies to track 1. EOS in 1 year 2. EOS in 2 years 3. and so on
Additionally on the topic of characteristics of supported software, I am currently using aging in my policies as an informative characteristic. I track that each component recent release is 1 years old, 2 years old, 3 years old, and so on to determine a priority for a deeper inspection of those aging components. 
Component current age, historic aging, and a declared EOS are all characteristics that can result in a DECISION. The facts supporting decisions are what counts with project/product leaders.

**Comment by Chriz76** [comment]
Thank you @nscuro and @ptdropper . 

Currently DT does not natively support eol and eos. So I am looking for a way to automatically integrate it.

I understand that you suggest not to use the private vulnerability repo workaround. So you think custom properties are better?

**Comment by daemon-byte** [comment]
Why not use the api to add and eol but allow it to be overridden for those usecases of extended support? I do like the idea of using them in policies though

**Comment by Chriz76** [comment]
After investigating the different possibilities I created a [PoC for an endoflife.date integration](https://github.com/Chriz76/endoflife-dependencytrack) that creates internal vulnerabilities. I am curious what you are thinking.

<img width="1013" height="360" alt="Image" src="https://github.com/user-attachments/assets/6fe8f5d7-2fcb-4d93-9aaa-d6b53c109d2e" />

**Comment by valentijnscholten** [comment]
The life cycle status, EOL date, etc. of a component version are  properties of a component(+version) so it should be modeled somewhere around that I would say.

**Comment by nscuro** [comment]
Had a look at the CLE spec (https://ecma-international.org/publications-and-standards/standards/ecma-428/), noting a few thoughts here:

* We should align our data model with CLE, i.e. storing event streams as-is.
* Data sources like endoflife.date, xeol could be mapped to CLE events.
* The data model should account for multiple sources. I fully expect that CLE adoption will be slow, and we'll need "adapters" (see above) in the mid-term. Eventually, when vendors publish 1st party CLE data, we must be able to differentiate them. Presumably, users will prefer 1st party data when available.
* The spec uses PURL identifiers but does not specify what role PURL qualifiers play. Say a CLE advertises `pkg:npm/acme-lib@1.2.3?foo=bar`, should that apply to *only* `pkg:npm/acme-lib@1.2.3?foo=bar`, or should variations like `pkg:npm/acme-lib@1.2.3` or `pkg:npm/acme-lib@1.2.3?foo=bar#baz` match as well? Essentially, the component identity problem is up to the consumer to figure out.
* The spec uses VERS for version ranges. Which is great, but it means ranges cannot be evaluated in the database, so JOINs are off the table (pun intended). Because many (all?) events in CLE have version ranges attached, determining the final state of a given component requires walking the entire event stream in-memory.
* It's highly likely we'll need an asynchronous job that computes the lifecycle status for all unique PURLs(?) in the portfolio. Which makes it all the more important the we store the original CLE streams.

Separately, I spotted @Chriz76's comments in https://github.com/endoflife-date/endoflife.date/issues/763#issuecomment-4036768078. It seems the PURL coverage of endoflife.date data is still very poor.

**Comment by habazie** [comment]
I think it would be great to move this forward and integrate EOL/EOS detection directly into Dependency-Track. The PoC prepared by Chriz76 demonstrates that this is feasible and could provide real value by identifying components that are no longer supported, not just outdated.

Having EOL information as a native Dependency-Track feature would be a very useful addition from both security and dependency lifecycle management perspectives.

**Comment by Chriz76** [comment]
> I think it would be great to move this forward and integrate EOL/EOS detection directly into Dependency-Track. The PoC prepared by Chriz76 demonstrates that this is feasible and could provide real value by identifying components that are no longer supported, not just outdated.
> 
> Having EOL information as a native Dependency-Track feature would be a very useful addition from both security and dependency lifecycle management perspectives.

Here my takeaways

Key Takeaways & Insights:

Integration: The core concept and technical integration of reporting End-of-Life (EoL) status as internal vulnerabilities directly into Dependency-Track works very well.

Data Challenges: The main bottleneck is the underlying data. Sources which cover only a fraction of existing libraries, and identifier mappings (PURL/CPE) are often incomplete or conflict with Linux distribution backports.

Recommended Strategy: Since there is currently no reliable, fully automated EoL database for the entire ecosystem, the most effective strategy for library consumers is to consistently and promptly update dependencies to their latest version.

See for details: https://github.com/Chriz76/endoflife-dependencytrack/issues/1#issuecomment-4803539327

**Comment by dcentrica** [comment]
@habazie, in the meantime [Metaport CE](https://gitlab.com/dcentrica/metaport/metaport-server) is OSS and provides EOL data (see screenshot ^). It integrates with DependencyTrack so end-users manage EOL and Vulns in the same system. Vuln prioritization itself is limited but is [earmarked for enrichment with EPSS and KEV](https://gitlab.com/dcentrica/metaport/metaport-server/-/work_items/461).
